gbasis/density.py

```diff
@@ -0,0 +1,44 @@
+"""Density Evaluation."""
+import numpy as np
+
+
+def eval_density(charge_density, cart_cont_orb):
+    """Return the evaluation of the density at different points.
+
+    Parameters
+    ----------
+    charge_density : np.ndarray(L,L,)
+        Charge density matrix.
+    cart_cont_orb : np.ndarray(L,N,)
+        Array of cartesian contracted orbitals at different grid points.
+
+    Returns
+    -------
+    density : np.ndarray(N,)
+        Density value at different grid points.
+
+    Raises
+    ------
+    TypeError
+        If cart_cont_orb is not an array of floats.
+        If charge_density is not an array of floats.
+    ValueError
+        If charge_density is not a 2-dimensional matrix.
+
+    """
+    # test that inputs have the correct shape and type
+    if not (isinstance(cart_cont_orb, np.ndarray) and cart_cont_orb.dtype == float):
+        raise TypeError("Contracted cartesian orbitals must be a numpy array of data type float.")
+    if not (isinstance(charge_density, np.ndarray) and charge_density.dtype == float):
+        raise TypeError("Charge density matrix must be a numpy array of data type float.")
+    if (len(charge_density.shape) != 2 or len(charge_density) != len(charge_density[0])):
+        raise ValueError("Charge density matrix must be a square matrix.")
+
+    # evaluate partial density contributions
+    partial_density = (cart_cont_orb[np.newaxis, :] * cart_cont_orb[:, np.newaxis]
+                       * charge_density[:, :, np.newaxis])
```

Comment:
Numpy, by default, broadcasts the matching shapes at the end of the dimensions i.e. shape(2, 3) and shape(1, 2, 3) can be broadcast but not shape(1, 2) and shape(1, 2, 3). This means that this code will break when the `cart_cont_orb` has a dimensionality greater than 2.

Here, you are hardcoding that the `cart_cont_orb` is 2 dimensional.

Reply:
Maybe that's okay because your docstring specifies that `cart_cont_orb` should be 2 dimensional. The only other use case would have been the `(L, Nx, Ny, Nz)` case, but the last three dimensions can be condensed down into a single dimension fairly easily. Maybe it'll be alright to support multiple dimensions, but you will need to reshape the `cart_cont_orb` into a two-dimensional array at the beginning and reshape to its original shape at the end.